rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,833 @@
+- **Task extraction vs Tuning**: Traditionally TVM has had three phases: i) Task extraction (find the fused sub-graphs
+  to tune), ii) Tuning (find a good schedule for those sub-graphs), and iii) Compilation (re-compile the model, now
+  retrieving schedules for all the anticipated sub-graphs from the cache.) However the Collage 'v2' prototype collapses
+  all these phases. This lets us lazily explore the implied search graph (nodes = partially rewritten models, edges =
+  selected of sub-graph and toolchain as a candidate kernel, cost = estimated sum of kernel costs plus launch penalties)
+  , and thus only pay the cost of tuning candidate kernels which could possibly influence the final partitioning.
+- **No non-local optimization**: Though Collage can explore the choice of sub-graph and toolchain, it cannot explore any
+  choices which require the arguments and/or result of the sub-graph to be rewritten. Thus Collage **cannot** be used to
+  search over:
+    - choice of layout for arguments/results (may require insertion of layout transforms),
+    - choice of memory scope for arguments/results (may require insertion of device copies),
+    - choice of device on which to host the kernel (ditto)
+      since all those choices can require changes beyond the candidates sub-graph.
+- the choice of layout for a kernel since any choice other than the model's default must be
+  'corrected' for by the inserted layout transformations. To support this efficiently we'd need to abandon the
+  simple-minded but fast `SubGraph` representation we describe below in favor of something like an EGraph
+  representation, which seems like a very large change for TVM.
+- **Dependency management**: Currently BYOC integrations tend to assume they are the only non-TVM toolchain in use. So
+  it's possible two toolchains introduce runtime dependencies which can't be satisfied. Collage has no notion of
+  dependencies or incompatibilities and may attemt to mix candidate kernels we can't support in prod. It's also possible
+  for two BYOC integrations to have incompatible runtimes.
+- **Additive kernel cost assumption**: Collage as per this design assumes the cost of running candidate kernels is
+  additive, plus a small launch penalty. However cache effects can dominate measured latency, particularly for 'light'
+  kernels. Thus there may be a **additive error** in the final result:
+
+  > additive_error = measured_latency(collage_partitioning) - sum_{kernel} (estimated_latency(kernel) + penalty)
+
+  The evolutionary search explored by the Collage paper can help here since it uses measured end-to-end model latency as
+  its cost function, but we're deferring that to future work.
+
+- **Limited search space**: Naively exploring all sub-graphs is O(n!), so we need to constrain the search. The easiest
+  approach is just to limit candidate kernels to sub-graphs of just a few operators. This can mean significatly faster
```

Comment:
This might be very bad on Hardware Accelerators that use extensive Fusion, e.g. EthosU. In these cases it would be probably be preferrable, to limit the search space by using a maximum number of splits of potentially fusible regions. Heuristics might be selected as an attribute of `CombineByPrimitivesFusionRule`.

Reply:
Though it's buried in a TODO I do intend to coalesce small candidates which ended up on the same target before continuing with the rest of the build. However that would be after collage has found what it thinks is the optimal partitioning, which has been measured w.r.t. these possibly too small partitions, and so may be too late. I'd like to experiment with the limits based on empirical measurements against common models and targets before speculating too far ahead here.

Reply:
Added section explaining why the 'partitioning' rules should, in a way, predict the downstream 'fusion' rules so as to not waste time partitioning across kernel boundaries while still keeping the maximum depth manageable.

I think we're still a bit speculative here since we're not yet deep into running 'v2' Collage on large models.